### PR TITLE
Improved runtime 12% by removing unnecessary deep copy.

### PR DESCRIPTION
**Pull request by PooyaHekmati**
Removed unnecessary deep copy from get_most_recent_date(); and added unit test for it. Also added type checker hints. This change reduced runtime from 173 seconds to 154 seconds on my laptop.

p.s. As a common practice, I applied auto-formatting to the changed files.  This changed lots of white spaces and make git identify many lines as changes. The only function that has changed is get_most_recent_date().

### CODE REVIEW

File: RUFAS/routines/animal/life_cycle/animal_events.py [hunk 1]
```diff
@@ -7,64 +7,71 @@
...
-		"""
-		dates = copy.deepcopy(list(self.events.keys()))
-		dates.reverse()
-		for date in dates:
-			if event_description in self.events[date]:
-				return date
-		return -1
+                Returns
+                -------
+                        the most recent age at which event_description, -1 if not found.
```
Comment (Fjordrunner) [inline RUFAS/routines/animal/life_cycle/animal_events.py]: Wouldn't we want the comment to read "at which event_description happened," or are we referring to that as a boolean in the comment?

File: RUFAS/routines/animal/life_cycle/animal_events.py [hunk 2]
```diff
@@ -7,64 +7,71 @@
...
+        """
+        dates = list(self.events.keys())
+        for i in range(-1, -len(dates)-1, -1):
```
Comment (Fjordrunner) [inline RUFAS/routines/animal/life_cycle/animal_events.py:75]: Out of curiosity, what is the reason that you're using negative values in the range() function here? I'm assuming it's to iterate through the list backwards? I just haven't seen this convention before, seems to work though!

**Review by Fjordrunner — APPROVED** [review]

**Comment by ltp23** [comment]
Thank you Pooya for making this pull request!

### Elapsed Time

I ran the program using the cProfiler module to measure some performance statistics for the master branch and this improvement branch to get a comparison. I also used the snakeviz module to visualize the results, and only the names of those functions with significant runtime are displayed by default. 

Example commands:
`$ python -m cProfile -o rufas-master.prof main.py`
`$ snakeviz rufas-master.prof`

Results from the master branch

(note the gray rectangle near lower left corner)

Master branch

Results from this improvement branch

Improvement branch

I see the function `get_most_recent_date()` is about 10x faster (though the exact numbers vary between computers). 

I also tried another profiler called yappi, which can measure the cpu time instead of just the elapsed or wall time, but it took much longer to perform a complete scan. 

### Linear search

I see you use the reverse linear scan to find the most recent date. I did experiment with another approach that uses reverse mapping and max heaps.  

Linear search: 
`add_event()`: O(1) time
`get_most_recent_date()`: O(N) time
Space: O(N) overall
The `events` dictionary: key = day, value = list of event descriptions

day_to_event_dict

Max heaps:
`add_event()`: O(logN) time
`get_most_recent_date()`: O(1) time for peeking
Space: O(2N) => O(N) overall. O(2N) because I had to build the following `events_reverse_dict` in addition to the `events` dictionary above. 

The `events_reverse_dict` dictionary: key = an event description, value = list of days, where the first element is guaranteed to be the most recent day. Because the `heapq` module in Python only supports min heaps so I had to negate all the numbers when adding and peeking. 
event_to_day_heap_dict

I also compared the runtime profiles between these two approaches. For a 2-year-long simulation, the max heap approach is about 5x faster and for a 6-year-long simulation about 9x faster. I think it may worth a try for much longer simulation. Other wise, linear scan is more straightforward and intuitive. 

### Testing

Here is an alternative version that uses parameters. 

```
@pytest.mark.parametrize(
    "events_list, event_descriptions, expected_days",
    [  # Each tuple is a scenario
        (
            [],  # events_list
            ['dummy'],  # event_descriptions
            [-1]  # expected_days
        ),
        (
            [(1, 2, 'event1'), (3, 4, 'event2'), (5, 6, 'event1'), (7, 8, 'event3')],
            ['event1', 'event2', 'event3', 'event0'],
            [5, 3, 7, -1]
        )
    ],
)
def test_get_most_recent_date2(events_list, event_descriptions, expected_days):
    """Unit test for function get_most_recent_date in file routines/animal/life_cycle/animal_events.py"""
    animal_event = AnimalEvents()
    for animal_age, simulation_day, event_description in events_list:
        animal_event.add_event(animal_age, simulation_day, event_description)

    for event_description, expected_day in zip(event_descriptions, expected_days):
        assert animal_event.get_most_recent_date(event_description) == expected_day
```

**Review by ltp23 — APPROVED** [review]

File: RUFAS/routines/animal/life_cycle/animal_events.py (hunk 2 above)
Comment (ancheypeng) [reply RUFAS/routines/animal/life_cycle/animal_events.py:75]: It seems like in the original code, the dates are revered before iterating, and using negative values in python range lets you iterate backwards creating the same effect without having to reverse the dates directly.

File: RUFAS/routines/animal/life_cycle/animal_events.py (hunk 1 above)
Comment (ancheypeng) [reply RUFAS/routines/animal/life_cycle/animal_events.py]: I agree, this might be a typo?

**Review by ancheypeng — APPROVED** [review]
Looks good!

**Comment by PooyaHekmati** [comment]
> Thank you Pooya for making this pull request!
> 
> ### Elapsed Time
> I ran the program using the cProfiler module to measure some performance statistics for the master branch and this improvement branch to get a comparison. I also used the snakeviz module to visualize the results, and only the names of those functions with significant runtime are displayed by default.
> 
> Example commands: `$ python -m cProfile -o rufas-master.prof main.py` `$ snakeviz rufas-master.prof`
> 
> Results from the master branch
> 
> (note the gray rectangle near lower left corner)
> 
> Master branch
> 
> Results from this improvement branch
> 
> Improvement branch
> 
> I see the function `get_most_recent_date()` is about 10x faster (though the exact numbers vary between computers).
> 
> I also tried another profiler called yappi, which can measure the cpu time instead of just the elapsed or wall time, but it took much longer to perform a complete scan.
> 
> ### Linear search
> I see you use the reverse linear scan to find the most recent date. I did experiment with another approach that uses reverse mapping and max heaps.
> 
> Linear search: `add_event()`: O(1) time `get_most_recent_date()`: O(N) time Space: O(N) overall The `events` dictionary: key = day, value = list of event descriptions
> 
> day_to_event_dict
> 
> Max heaps: `add_event()`: O(logN) time `get_most_recent_date()`: O(1) time for peeking Space: O(2N) => O(N) overall. O(2N) because I had to build the following `events_reverse_dict` in addition to the `events` dictionary above.
> 
> The `events_reverse_dict` dictionary: key = an event description, value = list of days, where the first element is guaranteed to be the most recent day. Because the `heapq` module in Python only supports min heaps so I had to negate all the numbers when adding and peeking. event_to_day_heap_dict
> 
> I also compared the runtime profiles between these two approaches. For a 2-year-long simulation, the max heap approach is about 5x faster and for a 6-year-long simulation about 9x faster. I think it may worth a try for much longer simulation. Other wise, linear scan is more straightforward and intuitive.
> 
> ### Testing
> Here is an alternative version that uses parameters.
> 
> ```
> @pytest.mark.parametrize(
>     "events_list, event_descriptions, expected_days",
>     [  # Each tuple is a scenario
>         (
>             [],  # events_list
>             ['dummy'],  # event_descriptions
>             [-1]  # expected_days
>         ),
>         (
>             [(1, 2, 'event1'), (3, 4, 'event2'), (5, 6, 'event1'), (7, 8, 'event3')],
>             ['event1', 'event2', 'event3', 'event0'],
>             [5, 3, 7, -1]
>         )
>     ],
> )
> def test_get_most_recent_date2(events_list, event_descriptions, expected_days):
>     """Unit test for function get_most_recent_date in file routines/animal/life_cycle/animal_events.py"""
>     animal_event = AnimalEvents()
>     for animal_age, simulation_day, event_description in events_list:
>         animal_event.add_event(animal_age, simulation_day, event_description)
> 
>     for event_description, expected_day in zip(event_descriptions, expected_days):
>         assert animal_event.get_most_recent_date(event_description) == expected_day
> ```

We discussed this offline. Performing the search with better time complexity of O(N) requires increasing the space complexity from O(1) to O (N) which sounds risky as we are not sure about the size of the real data; and, O(N) for time complexity is acceptable.

**Pull request merged**